File: app/lib/core/e2ee/identity_trust.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';

const _trustRegistryVersion = 1;
const _securityCodeDomain = 'redcode-im/e2ee/security-code/v1\u0000';

class E2eeRootIdentity {
  const E2eeRootIdentity({
    required this.userId,
    required this.publicKey,
    required this.fingerprint,
    required this.protocolVersion,
  });

  final String userId;
  final Uint8List publicKey;
  final Uint8List fingerprint;
  final int protocolVersion;

  Map<String, Object> toJson() => {
    'user_id': userId,
    'public_key': base64Encode(publicKey),
    'fingerprint': base64Encode(fingerprint),
    'protocol_version': protocolVersion,
  };

  factory E2eeRootIdentity.fromJson(Map<String, dynamic> json) {
    return E2eeRootIdentity(
      userId: json['user_id'] as String,
      publicKey: base64Decode(json['public_key'] as String),
      fingerprint: base64Decode(json['fingerprint'] as String),
      protocolVersion: json['protocol_version'] as int,
    );
  }
}

class E2eeIdentityTrustRecord {
  const E2eeIdentityTrustRecord({
    required this.trusted,
    required this.trustedAt,
    this.pending,
    this.changedAt,
  });

  final E2eeRootIdentity trusted;
  final DateTime trustedAt;
  final E2eeRootIdentity? pending;
  final DateTime? changedAt;

  bool get isChanged => pending != null;

  Map<String, Object?> toJson() => {
    'trusted': trusted.toJson(),
    'trusted_at': trustedAt.toUtc().toIso8601String(),
    'pending': pending?.toJson(),
    'changed_at': changedAt?.toUtc().toIso8601String(),
  };

  factory E2eeIdentityTrustRecord.fromJson(Map<String, dynamic> json) {
    final pending = json['pending'];
    return E2eeIdentityTrustRecord(
      trusted: E2eeRootIdentity.fromJson(
        (json['trusted'] as Map).cast<String, dynamic>(),
      ),
      trustedAt: DateTime.parse(json['trusted_at'] as String).toUtc(),
      pending: pending == null
          ? null
          : E2eeRootIdentity.fromJson((pending as Map).cast<String, dynamic>()),
      changedAt: json['changed_at'] == null
          ? null
          : DateTime.parse(json['changed_at'] as String).toUtc(),
    );
  }
}

enum E2eeIdentityTrustDecision { firstUseTrusted, trusted, identityChanged }

class E2eeIdentityNotTrustedException implements Exception {
  const E2eeIdentityNotTrustedException(this.message);

  final String message;

  @override
  String toString() => message;
}

abstract interface class E2eeIdentityTrustStore {
  Future<Map<String, E2eeIdentityTrustRecord>> readRecords(String accountId);
  Future<void> writeRecords(
    String accountId,
    Map<String, E2eeIdentityTrustRecord> records,
  );
  Future<void> deleteRecords(String accountId);
}

class E2eeIdentityTrustManager {
  E2eeIdentityTrustManager({
    required E2eeIdentityTrustStore store,
    DateTime Function()? now,
  }) : _store = store,
       _now = now ?? DateTime.now;

  final E2eeIdentityTrustStore _store;
  final DateTime Function() _now;

  Future<E2eeIdentityTrustDecision> observe(
    String accountId,
    E2eeRootIdentity identity,
  ) async {
    _validateIdentity(identity);
    final records = await _store.readRecords(accountId);
    final existing = records[identity.userId];
    if (existing == null) {
      records[identity.userId] = E2eeIdentityTrustRecord(
        trusted: identity,
        trustedAt: _now().toUtc(),
      );
      await _store.writeRecords(accountId, records);
      return E2eeIdentityTrustDecision.firstUseTrusted;
    }
    if (_sameIdentity(existing.trusted, identity)) {
      return E2eeIdentityTrustDecision.trusted;
    }
    records[identity.userId] = E2eeIdentityTrustRecord(
      trusted: existing.trusted,
      trustedAt: existing.trustedAt,
      pending: identity,
      changedAt: existing.changedAt ?? _now().toUtc(),
    );
    await _store.writeRecords(accountId, records);
    return E2eeIdentityTrustDecision.identityChanged;
  }

  Future<E2eeIdentityTrustRecord?> record(
    String accountId,
    String targetUserId,
  ) async => (await _store.readRecords(accountId))[targetUserId];

  Future<E2eeIdentityTrustRecord> requireTrusted(
    String accountId,
    String targetUserId,
  ) async {
    final existing = await record(accountId, targetUserId);
    if (existing == null) {
      throw const E2eeIdentityNotTrustedException('联系人 E2EE 身份尚未验证');
    }
    if (existing.isChanged) {
      throw const E2eeIdentityNotTrustedException('联系人 E2EE 身份已变化，核验安全码后才能发送');
    }
    return existing;
  }

  Future<E2eeIdentityTrustRecord> retrust(
    String accountId,
    String targetUserId,
  ) async {
    final records = await _store.readRecords(accountId);
    final existing = records[targetUserId];
    if (existing?.pending == null) {
      throw StateError('没有待确认的 E2EE 身份变化');
    }
    final accepted = E2eeIdentityTrustRecord(
      trusted: existing!.pending!,
      trustedAt: _now().toUtc(),
    );
    records[targetUserId] = accepted;
    await _store.writeRecords(accountId, records);
    return accepted;
  }

  static String securityCode(E2eeRootIdentity first, E2eeRootIdentity second) {
    _validateIdentity(first);
    _validateIdentity(second);
    final identities = [first, second]
      ..sort((left, right) => left.userId.compareTo(right.userId));
    final input = BytesBuilder(copy: false)
      ..add(utf8.encode(_securityCodeDomain));
    for (final identity in identities) {
      _addLengthPrefixed(input, utf8.encode(identity.userId));
      _addLengthPrefixed(input, identity.fingerprint);
    }
    final hex = sha256.convert(input.takeBytes()).toString().toUpperCase();
    return [
      for (var i = 0; i < hex.length; i += 4) hex.substring(i, i + 4),
    ].join(' ');
  }

  static Uint8List encodeRegistry(
    Map<String, E2eeIdentityTrustRecord> records,
  ) {
    final sorted = records.keys.toList()..sort();
    return Uint8List.fromList(
      utf8.encode(
        jsonEncode({
          'version': _trustRegistryVersion,
          'records': {for (final key in sorted) key: records[key]!.toJson()},
        }),
      ),
    );
  }

  static Map<String, E2eeIdentityTrustRecord> decodeRegistry(List<int> data) {
    final decoded = jsonDecode(utf8.decode(data));
    if (decoded is! Map<String, dynamic> ||
        decoded['version'] != _trustRegistryVersion ||
        decoded['records'] is! Map) {
      throw const FormatException('E2EE 信任记录格式无效');
    }
    return (decoded['records'] as Map).map(
      (key, value) => MapEntry(
        key as String,
        E2eeIdentityTrustRecord.fromJson(
          (value as Map).cast<String, dynamic>(),
        ),
      ),
    );
  }

  static bool _sameIdentity(E2eeRootIdentity left, E2eeRootIdentity right) =>
      left.userId == right.userId &&
      left.protocolVersion == right.protocolVersion &&
      _bytesEqual(left.publicKey, right.publicKey) &&
      _bytesEqual(left.fingerprint, right.fingerprint);

  static void _validateIdentity(E2eeRootIdentity identity) {
    if (identity.userId.trim().isEmpty ||
        identity.publicKey.isEmpty ||
        identity.fingerprint.length < 16 ||
        identity.protocolVersion != 1) {
      throw const FormatException('E2EE 根身份无效');
    }
  }

  static bool _bytesEqual(List<int> left, List<int> right) {
    if (left.length != right.length) return false;
    var difference = 0;
    for (var i = 0; i < left.length; i++) {
      difference |= left[i] ^ right[i];
    }
    return difference == 0;
  }

  static void _addLengthPrefixed(BytesBuilder output, List<int> value) {
    if (value.length > 0xffff) throw const FormatException('安全码字段过长');
    output.add([(value.length >> 8) & 0xff, value.length & 0xff]);
    output.add(value);
  }
}

File: h5-app/src/e2ee/identity-trust.ts
const REGISTRY_VERSION = 1;
const SECURITY_CODE_DOMAIN = 'redcode-im/e2ee/security-code/v1\0';

export interface E2eeRootIdentity {
  userId: string;
  publicKey: Uint8Array;
  fingerprint: Uint8Array;
  protocolVersion: number;
}

export interface E2eeIdentityTrustRecord {
  trusted: E2eeRootIdentity;
  trustedAt: string;
  pending?: E2eeRootIdentity;
  changedAt?: string;
}

export type E2eeIdentityTrustDecision = 'first-use-trusted' | 'trusted' | 'identity-changed';

export class E2eeIdentityNotTrustedError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'E2eeIdentityNotTrustedError';
  }
}

export interface E2eeIdentityTrustStore {
  readRecords(accountId: string): Promise<Record<string, E2eeIdentityTrustRecord>>;
  writeRecords(accountId: string, records: Record<string, E2eeIdentityTrustRecord>): Promise<void>;
  deleteRecords(accountId: string): Promise<void>;
}

export class E2eeIdentityTrustManager {
  constructor(
    private readonly store: E2eeIdentityTrustStore,
    private readonly now: () => Date = () => new Date(),
  ) {}

  async observe(accountId: string, identity: E2eeRootIdentity): Promise<E2eeIdentityTrustDecision> {
    validateIdentity(identity);
    const records = await this.store.readRecords(accountId);
    const existing = records[identity.userId];
    if (!existing) {
      records[identity.userId] = { trusted: identity, trustedAt: this.now().toISOString() };
      await this.store.writeRecords(accountId, records);
      return 'first-use-trusted';
    }
    if (sameIdentity(existing.trusted, identity)) return 'trusted';
    records[identity.userId] = {
      trusted: existing.trusted,
      trustedAt: existing.trustedAt,
      pending: identity,
      changedAt: existing.changedAt ?? this.now().toISOString(),
    };
    await this.store.writeRecords(accountId, records);
    return 'identity-changed';
  }

  async record(accountId: string, targetUserId: string): Promise<E2eeIdentityTrustRecord | undefined> {
    return (await this.store.readRecords(accountId))[targetUserId];
  }

  async requireTrusted(accountId: string, targetUserId: string): Promise<E2eeIdentityTrustRecord> {
    const existing = await this.record(accountId, targetUserId);
    if (!existing) throw new E2eeIdentityNotTrustedError('联系人 E2EE 身份尚未验证');
    if (existing.pending) {
      throw new E2eeIdentityNotTrustedError('联系人 E2EE 身份已变化，核验安全码后才能发送');
    }
    return existing;
  }

  async retrust(accountId: string, targetUserId: string): Promise<E2eeIdentityTrustRecord> {
    const records = await this.store.readRecords(accountId);
    const existing = records[targetUserId];
    if (!existing?.pending) throw new Error('没有待确认的 E2EE 身份变化');
    const accepted = { trusted: existing.pending, trustedAt: this.now().toISOString() };
    records[targetUserId] = accepted;
    await this.store.writeRecords(accountId, records);
    return accepted;
  }

  static encodeRegistry(records: Record<string, E2eeIdentityTrustRecord>): Uint8Array {
    const serialized = Object.fromEntries(
      Object.keys(records).sort().map((key) => [key, serializeRecord(records[key]!)]),
    );
    return new TextEncoder().encode(JSON.stringify({ version: REGISTRY_VERSION, records: serialized }));
  }

  static decodeRegistry(data: Uint8Array): Record<string, E2eeIdentityTrustRecord> {
    const decoded = JSON.parse(new TextDecoder('utf-8', { fatal: true }).decode(data)) as {
      version?: unknown;
      records?: unknown;
    };
    if (decoded.version !== REGISTRY_VERSION || !isRecord(decoded.records)) {
      throw new Error('E2EE 信任记录格式无效');
    }
    return Object.fromEntries(
      Object.entries(decoded.records).map(([key, value]) => [key, parseRecord(value)]),
    );
  }

  static async securityCode(
    first: E2eeRootIdentity,
    second: E2eeRootIdentity,
    cryptoProvider: Crypto = globalThis.crypto,
  ): Promise<string> {
    validateIdentity(first);
    validateIdentity(second);
    if (!cryptoProvider?.subtle) throw new Error('WebCrypto 不可用');
    const identities = [first, second].sort((left, right) => (
      left.userId < right.userId ? -1 : left.userId > right.userId ? 1 : 0
    ));
    const parts: Uint8Array[] = [new TextEncoder().encode(SECURITY_CODE_DOMAIN)];
    for (const identity of identities) {
      parts.push(lengthPrefixed(new TextEncoder().encode(identity.userId)));
      parts.push(lengthPrefixed(identity.fingerprint));
    }
    const digest = new Uint8Array(await cryptoProvider.subtle.digest('SHA-256', concat(parts)));
    const hex = Array.from(digest, (byte) => byte.toString(16).padStart(2, '0')).join('').toUpperCase();
    return hex.match(/.{4}/g)!.join(' ');
  }
}

const serializeIdentity = (identity: E2eeRootIdentity) => ({
  user_id: identity.userId,
  public_key: bytesToBase64(identity.publicKey),
  fingerprint: bytesToBase64(identity.fingerprint),
  protocol_version: identity.protocolVersion,
});

const parseIdentity = (value: unknown): E2eeRootIdentity => {
  if (!isRecord(value)) throw new Error('E2EE 根身份格式无效');
  const identity = {
    userId: String(value.user_id ?? ''),
    publicKey: base64ToBytes(String(value.public_key ?? '')),
    fingerprint: base64ToBytes(String(value.fingerprint ?? '')),
    protocolVersion: Number(value.protocol_version),
  };
  validateIdentity(identity);
  return identity;
};

const serializeRecord = (record: E2eeIdentityTrustRecord) => ({
  trusted: serializeIdentity(record.trusted),
  trusted_at: record.trustedAt,
  pending: record.pending ? serializeIdentity(record.pending) : null,
  changed_at: record.changedAt ?? null,
});

const parseRecord = (value: unknown): E2eeIdentityTrustRecord => {
  if (!isRecord(value) || typeof value.trusted_at !== 'string') {
    throw new Error('E2EE 信任记录格式无效');
  }
  return {
    trusted: parseIdentity(value.trusted),
    trustedAt: value.trusted_at,
    pending: value.pending == null ? undefined : parseIdentity(value.pending),
    changedAt: typeof value.changed_at === 'string' ? value.changed_at : undefined,
  };
};

const validateIdentity = (identity: E2eeRootIdentity) => {
  if (!identity.userId.trim() || !identity.publicKey.length || identity.fingerprint.length < 16 || identity.protocolVersion !== 1) {
    throw new Error('E2EE 根身份无效');
  }
};

const sameIdentity = (left: E2eeRootIdentity, right: E2eeRootIdentity) =>
  left.userId === right.userId
  && left.protocolVersion === right.protocolVersion
  && bytesEqual(left.publicKey, right.publicKey)
  && bytesEqual(left.fingerprint, right.fingerprint);

const bytesEqual = (left: Uint8Array, right: Uint8Array) => {
  if (left.length !== right.length) return false;
  let difference = 0;
  for (let index = 0; index < left.length; index += 1) difference |= left[index]! ^ right[index]!;
  return difference === 0;
};

const lengthPrefixed = (value: Uint8Array) => {
  if (value.length > 0xffff) throw new Error('安全码字段过长');
  return concat([new Uint8Array([value.length >> 8, value.length & 0xff]), value]);
};

const concat = (parts: Uint8Array[]) => {
  const output = new Uint8Array(parts.reduce((sum, item) => sum + item.length, 0));
  let offset = 0;
  for (const part of parts) {
    output.set(part, offset);
    offset += part.length;
  }
  return output;
};

const bytesToBase64 = (value: Uint8Array) => btoa(String.fromCharCode(...value));
const base64ToBytes = (value: string) => Uint8Array.from(atob(value), (char) => char.charCodeAt(0));
const isRecord = (value: unknown): value is Record<string, unknown> => Boolean(value && typeof value === 'object' && !Array.isArray(value));
